hazelcast-jclouds/src/main/java/com/hazelcast/jclouds/ComputeServiceBuilder.java

```diff
@@ -0,0 +1,279 @@
+            throw new UnsupportedOperationException("Both credential and credentialPath are set. Use only one method.");
+        }
+
+        if (credentialPath != null) {
+            credential = getCredentialFromFile(credential, credentialPath);
+        }
+
+        if (LOGGER.isFinestEnabled()) {
+            LOGGER.finest("Using CLOUD_PROVIDER: " + cloudProvider);
+        }
+
+        ContextBuilder contextBuilder = newContextBuilder(cloudProvider);
+
+        Properties jcloudsProperties = buildRegionZonesConfig();
+        buildTagConfig();
+        buildNodeFilter();
+
+        computeService = contextBuilder.overrides(jcloudsProperties)
+                            .credentials(identity, credential)
+                            .buildView(ComputeServiceContext.class)
+                            .getComputeService();
+
+        return computeService;
+    }
+
+    private Properties buildRegionZonesConfig() {
+        final String regions = getOrNull(JCloudsProperties.REGIONS);
+        final String zones = getOrNull(JCloudsProperties.ZONES);
+        Properties jcloudsProperties = newOverrideProperties();
+        if (regions != null) {
+            jcloudsProperties.setProperty(LocationConstants.PROPERTY_REGIONS, regions);
+            tokenizeValues(regions, regionsSet);
+        }
+        if (zones != null) {
+            jcloudsProperties.setProperty(LocationConstants.PROPERTY_ZONES, zones);
+            jcloudsProperties.setProperty(LocationConstants.PROPERTY_REGIONS, regions);
+            tokenizeValues(zones, zonesSet);
+        }
+        return jcloudsProperties;
+    }
+
+    private void buildTagConfig() {
+        final String tagKeys = getOrNull(JCloudsProperties.TAG_KEYS);
+        final String tagValues = getOrNull(JCloudsProperties.TAG_VALUES);
+        if (tagKeys != null && tagValues != null) {
+            tokenizeValues(tagKeys, tagsKeysSet);
+            tokenizeValues(tagValues, tagsValuesSet);
+            if (tagsKeysSet.size() != tagsValuesSet.size()) {
+                throw new InvalidConfigurationException("Tags keys and value count does not match.");
+            }
+        }
+    }
+
+    private void tokenizeValues(String params, Set<String> paramSet) {
+        StringTokenizer st = new StringTokenizer(params, ",");
+        while (st.hasMoreTokens()) {
+            String token = st.nextToken();
+            paramSet.add(token.trim());
+        }
+    }
+
+    private void buildNodeFilter() {
+
+        nodesFilter = new Predicate<ComputeMetadata>() {
+             @Override
+             public boolean apply(ComputeMetadata nodeMetadata) {
+                if (nodeMetadata == null) {
+                    return false;
+                }
+                for (String key: tagsKeysSet) {
+                    for (String value: tagsValuesSet) {
+                        if (!nodeMetadata.getUserMetadata().get(key).equals(value)) {
+                            return false;
```

Comment:
@bilalyasar Thanks for pointing out that. I did not test for multiple key value pairs, but this implementation would cause problems. I create a pair for each (tag-key,tag-value) tuple with `AbstractMap.SimpleImmutableEntry ` and make checks that way.